snapshotter/demux/metrics/proxy.go

```diff
@@ -30,10 +32,71 @@ const (
 	MaxMetricsResponseSize = 32768 // 32 KB
 )
 
+// Monitor is used for port and lifecycle management of metrics proxies.
+type Monitor struct {
+	// monitorChan indicates a port is free for metrics proxy server.
+	ch chan int
+	// portMap tracks used and unused metrics proxy server ports.
+	portMap map[int]bool
+	// startPort is first port in the configurable range of ports to serve metrics proxies from.
+	startPort int
+	// endPort is last port in the configurable range of ports to serve metrics proxies from.
+	endPort int
+	// mu is the lock used for threadsafe reading and writing of the metrics proxy port map.
+	mu *sync.Mutex
+}
+
+// NewMonitor returns a new metrics proxy monitor.
+func NewMonitor(startPort, endPort int) *Monitor {
+	return &Monitor{make(chan int), make(map[int]bool), startPort, endPort, &sync.Mutex{}}
+}
+
+// Start receives messages from the proxy server channel indiciating a server has closed or terminated.
+// It marks the metrics proxy server's port as available.
+func (m *Monitor) Start() error {
+	for freePort := range m.ch {
+		m.mu.Lock()
+		delete(m.portMap, freePort)
+		m.mu.Unlock()
+	}
+
+	return nil
+}
+
+// Stop closes the metrics proxy monitoring channel.
+func (m *Monitor) Stop() {
+	close(m.ch)
+}
+
+// findOpenPort returns a port for use by a metrics proxy server.
+func (m *Monitor) findOpenPort() (int, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	metricsProxyPort := -1
+	for p := m.startPort; p <= m.endPort; p++ {
+		if _, ok := m.portMap[p]; !ok {
+			metricsProxyPort = p
+			m.portMap[p] = true
+			break
+		}
+	}
+
+	if metricsProxyPort < 0 {
+		return 0, fmt.Errorf("invalid port: %d", metricsProxyPort)
+	}
+
+	return metricsProxyPort, nil
```

Comment:
```suggestion
    for port, ok := range m.portMap {
        if !ok {
            m.portMap[port] = true
            return port, nil
        }
    }
    
    return 0, fmt.Errorf("unable to allocate a port for metrics proxy server")
```

Reply:
I elected to iterate through the port range rather than the map elements because we are treating this map as a set. The portMap itself has no concept of the port range. So if we were to take this change and iterate through the map itself, we wouldn't actually be considering the port range.
I'll call that we treat it as a set in a comment